database/upload.go

```diff
@@ -80,10 +82,61 @@ func (db *DB) UploadsBySkylink(ctx context.Context, skylink Skylink, offset, pag
 	if err := validateOffsetPageSize(offset, pageSize); err != nil {
 		return nil, 0, err
 	}
-	matchStage := bson.D{{"$match", bson.D{{"skylink_id", skylink.ID}}}}
+	matchStage := bson.D{{"$match", bson.D{
+		{"skylink_id", skylink.ID},
+		{"unpinned", false},
+	}}}
 	return db.uploadsBy(ctx, matchStage, offset, pageSize)
 }
 
+// UnpinUpload unpins a single upload by this user. Returns the number of
+// unpinned uploads.
+func (db *DB) UnpinUpload(ctx context.Context, uploadId string, user User) (int64, error) {
```

Comment:
q: I don't really understand the difference between this and `unpinUploads`. One unpins by an upload ID and the other by a skylink? Shouldn't they amount to the same operation?

Reply:
One unpins a specific upload while the other unpins all uploads of the given skylink by the given user. The difference appears when the user uploads the exact same file multiple times. Those uploads will result in the same skylink.

Reply:
how do you unpin a specific upload? I thought you could only unpin by skylink, hence why I thought both operations would be the same.

Reply:
When a user uploads a file that results in a skyfile stored on Sia and an upload record in our database. The skyfile is posted on the network by the portal's Sia node. The upload records are not directly connected to the uploaded skyfiles, they are just a record we keep on which users are paying for the upkeep of the given skyfile. While there is at least one user paying for that upkeep, the portal's node will not unpin the file from the network. The portal's mechanism for monitoring this are the upload records in the DB. 

When the user chooses to unpin a single upload, we delete a single upload record they have for that skyfile. Or, rather, mark it as `unpinned`. When they choose to unpin all, we mark all of the uploads they have made for that skyfile as `unpinned`. If no other user is pinning this skyfile, the portal's node will unpin it from the network on its next cleanup cycle. If, however, other users are still pinning the skyfile the portal will keep it pinned.

The unpinning of a specific upload by its id was implemented mostly for the purposes of the front end, so a user clicking `unpin` on their second upload of a given skyfile will see that specific second upload as unpinned. This allows for smooth and predictable UX.